### PR TITLE
Optimized Scheduling

### PR DESCRIPTION
**Pull request by arjnpandey**

### CODE REVIEW

File: src/schedule.py [hunk 1]
```diff
@@ -3,9 +3,10 @@
 import networkx as nx
 import matplotlib.pyplot as plt
 import numpy as np
-
+import cvxpy as cp
```
Comment (fazerlicourice7) [inline src/schedule.py:6]: add a blank line below

File: src/schedule.py [hunk 2]
```diff
@@ -3,9 +3,10 @@
...
 from global_constants import SEED
 import sim_util
+import heapq
```
Comment (fazerlicourice7) [inline src/schedule.py:9]: move heapq above with other 3rd party libs

File: src/schedule.py [hunk 3]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+    TODO: Add exact hw schema constraints as resource constraints. Right now we assume hardware is
+    all to all.
+    """
+    constraints = []
+    vars = []
+    # graph nodes
+    id = 0
+    for node in graph_nodes:
+        curr_var = cp.Variable(2, name = node[0]) # first one is start time and last one is end time
+        # curr_var.name = node[0]
```
Comment (fazerlicourice7) [inline src/schedule.py:183]: del this

File: src/schedule.py [hunk 4]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+        vars.append(curr_var)
+        # start time + latency = end time for each operating node
+        node[1]['scheduling_id'] = id
+        id += 1
+        constraints.append(curr_var[0] >= 0)
+        # constraints.append(curr_var[1] >= 0)
```
Comment (fazerlicourice7) [inline src/schedule.py:189]: del this

File: src/schedule.py [hunk 5]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+        if 'cost' in node[1].keys():
+            constraints.append(curr_var[0] + node[1]['cost'] == curr_var[1])
+
+    # print(graph_nodes)
+    # nx.set_node_attributes(graph, graph_nodes)
```
Comment (fazerlicourice7) [inline src/schedule.py:194]: del this

File: src/schedule.py [hunk 6]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+    for u, v in graph.edges():
+        # if 'idx'
+        source_id = int(graph_nodes[u]['scheduling_id'])
+        dest_id = int(graph_nodes[v]['scheduling_id'])
+        constraints.append(vars[source_id][1] - vars[dest_id][0] <= 0.0)
+    
+    topological_order = longest_path_first_topological_sort(graph)
+    resource_constraints = []
+    for i in range(len(topological_order)):
+        # curr_reg_count = {'Regs': 0, 'Add': 0, 'Mult': 0, 'Buf': 0, 'Eq': 0, 'stall': 0}
```
Comment (fazerlicourice7) [inline src/schedule.py:205]: del this

File: src/schedule.py [hunk 7]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+        curr_func_count = 0
+        start_node = topological_order[i]
+        if graph.nodes[start_node]['function'] not in curr_reg_count:
+            continue
+        # curr_reg_count[graph.nodes[start_node]['function']] += 1
```
Comment (fazerlicourice7) [inline src/schedule.py:210]: del this

File: src/schedule.py [hunk 8]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+        curr_func_count += 1
+        for j in range(i + 1, len(topological_order)):
+            curr_node = topological_order[j]
+            if graph.nodes[curr_node]['function'] not in curr_reg_count.keys():
+                    continue
+            # curr_reg_count[graph.nodes[curr_node]['function']] += 1
```
Comment (fazerlicourice7) [inline src/schedule.py:216]: del this

File: src/schedule.py [hunk 9]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+            if graph.nodes[curr_node]['function'] == graph.nodes[start_node]['function']:
+                curr_func_count += 1
+                if curr_func_count > 2*hw_element_counts[graph.nodes[curr_node]['function']]:
+                    break
+                if curr_func_count > hw_element_counts[graph.nodes[curr_node]['function']]:
+                    # add a constraint
+                    resource_constraints.append(vars[graph.nodes[start_node]['scheduling_id']][0] - vars[graph.nodes[curr_node]['scheduling_id']][0] <= -graph.nodes[start_node]['cost'])
+                    # break
+        # print(curr_reg_count)
```
Comment (fazerlicourice7) [inline src/schedule.py:225]: del this

File: src/schedule.py [hunk 10]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
...
+    constraints += resource_constraints
+    obj = cp.Minimize(vars[graph_nodes['end']['scheduling_id']][0])
+    # obj = cp.Minimize(all_nodes)
```
Comment (fazerlicourice7) [inline src/schedule.py:228]: del

File: src/schedule.py [hunk 11]
```diff
@@ -103,7 +166,71 @@ def assign_upstream_path_lengths(graph):
     
     return graph
 
-def schedule(computation_graph, hw_element_counts):
+def schedule(graph, hw_element_counts):
```
Comment (fazerlicourice7) [inline src/schedule.py:169]: rename to sdc_schedule.
and call sdc_schedule from simulator and symbolic_simulator

**Review by fazerlicourice7 — CHANGES_REQUESTED** [review]

File: src/schedule.py (hunk 2 above)
Comment (arjnpandey) [reply src/schedule.py]: done

**Review by fazerlicourice7 — APPROVED** [review]

**Pull request merged**